src/dune_action_plugin/dune_action_plugin.mli

```diff
@@ -71,9 +72,21 @@ module V1 : sig
       listing of a [directory] and all source and target files contained in
       that directory. Computation will result in a directory listing.
 
-      BUG: [read_directory] doesn't work for empty directories. *)
+      BUG: [read_directory] doesn't work correctly for empty directories.
+
+      BUG: the returned listing includes directories even though that dependency
+      is not tracked.
+  *)
   val read_directory : path:Path.t -> string list t
 
+  (** [read_directory_with_glob ~path:directory ~glob] returns a computation
+      depending on a listing of a [directory] filtered by glob and resulting in
+      that listing.
```

Comment:
Might be worth mentioning that this is better than filtering the output of `read_directory` since the filtering is done on the Dune side. In fact, I'm wondering if this should completely replace `read_directory`, i.e. one would have to use the `*` glob in order to get everything. What do you think?

Reply:
Added to comment:

```
      This is better than filtering the output of `read_directory` because this
      makes dune aware of the filtering, so dune won't re-run the action when
      the directory changes in an unimportant way.
```

>  i.e. one would have to use the `*` glob in order to get everything

The * glob is strange because it doesn't match hidden files. We added `Glob.universal` to make things easier, but it still seemed that a simple primitive with no filtering is useful at least as a stepping stone to learn your way around the library.

Do you think it's not worth having after all?

Reply:
I'm just thinking that this looks like a reasonable way to write the code:

```ocaml
let+ l = read_directory ~path in
List.filter l ~f:(fun fn -> Filename.check_extension fn ".txt")
```

while this is much more dubious:

```ocaml
let+ l = read_directory ~path ~glob:Glob.universal in
List.filter l ~f:(fun fn -> Filename.check_extension fn ".txt")
```

So in a way we'd be encouraging users to do the right thing by giving them less choices, at the cost of a slightly higher learning curve.

Reply:
Yeah, that seems OK.
I pushed a commit that removes `read_directory`. Even though that's a non-backwards-compatible change, I suppose we didn't release V1 yet and it seems unlikely to break much.

Reply:
Indeed, there is no stability guarantee on master until it's released :)